source/utils/LogManager.h

```diff
@@ -18,45 +18,72 @@
 #ifndef LOGMANAGER_H
 #define LOGMANAGER_H
 
+#include <memory>
+#include <mutex>
 #include <string>
 
 #include <OgreSingleton.h>
 
-#define OD_STRING_S(x) #x
-#define OD_STRING_S_(x) OD_STRING_S(x)
-#define S__LINE__ OD_STRING_S_(__LINE__)
-#define OD_LOG_ERR(a)      LogManager::getSingleton().logMessage("ERROR: " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" ") + (a), LogMessageLevel::CRITICAL)
-#define OD_LOG_WRN(a)      LogManager::getSingleton().logMessage("WARNING: " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" ") + (a), LogMessageLevel::WARNING)
-#define OD_LOG_INF(a)      LogManager::getSingleton().logMessage(a, LogMessageLevel::NORMAL)
-#define OD_LOG_DBG(a)      LogManager::getSingleton().logMessage(a, LogMessageLevel::TRIVIAL)
-#define OD_ASSERT_TRUE(a)        if(!(a)) LogManager::getSingleton().logMessage("ERROR: Assert failed file " + std::string(__FILE__) + " line " + std::string(S__LINE__), LogMessageLevel::CRITICAL)
-#define OD_ASSERT_TRUE_MSG(a,b)  if(!(a)) LogManager::getSingleton().logMessage("ERROR: Assert failed file " + std::string(__FILE__) + " line " + std::string(S__LINE__) + std::string(" info : ") + b, LogMessageLevel::CRITICAL)
+#include "utils/Helper.h"
+#include "utils/LogMessageLevel.h"
+#include "utils/LogSink.h"
 
+#ifndef OD_LOG_ALLOW_DEPRECATED
+    #define OD_LOG_ALLOW_DEPRECATED 1
+#endif
 
-enum class LogMessageLevel
-{
-    TRIVIAL = 1,
-    NORMAL,
-    WARNING,
-    CRITICAL
-};
+#if OD_LOG_ALLOW_DEPRECATED
+    // DEPRECATED: Use OD_TRACE_ERR instead.
```

Comment:
@Bertram25 Yes.The reason I want to keep them is that I wanted to prevent creating a change that affected *all* files in the source tree. I agree with you that it's not ideal, however.

@hwoarangmy Unfortunately, the API isn't the same between the new and the old style. The old style creates an `std::string` that is passed to the `logMessage` function, while the new style uses variable arguments to create a `printf`-style logging interface.

I have written at least a dozen logging interfaces throughout my career, both using streams and using `printf`. There are clear advantages and disadvantages to both.

Streams:

    + Type-safety. A string will never be interpreted as a number.
    + Extensible. Easy to create a custom "operator <<" overload for a specific type, so you can just write "ss << "position: " << vec3(1.0f, 1.0, 1.0f);" for example.
    - Slow. Multiple dynamic memory allocations means they're not ideal for high-performance logging.
    - Cumbersome. Ever tried writing a hex decimal to a stringstream? "ss << "0x" << std::hex << std::width(8) << number;" Compare that to "sprintf_s(buffer, "0x%08x", number);".
    - Can't be compiled out. Logging interfaces tend to become cumbersome because logging can no longer be disabled using the preprocessor.

Printf:

    + Familiar to C users. If you know printf, you know what to expect when you see "LOG("Stuff %d", thangs);". If not, you can look up the syntax.
    + Compact format. Printf syntax is like a mini-language where you can specify exactly what the output should be.
    - No type-safety. Put a string in a "%d" and watch the fireworks!
    - Don't try to move arguments around. "LOG("(%d) (%s:%i) %s: %s", timestamp, filename, line, message, groceries");" No, wait, the filename should be at the end...

The best logging system I've ever seen is part of a AAA game engine. It uses a `printf` syntax, but the arguments are _numbered_. For example:

    LOG("What a {0} day!", m_AIMood);

This gives you the compactness of `printf`, the extensibility of `std::stringstream`, and type-safety! Unfortunately, in typical AAA game engine fashion, it's horribly complicated and only feasible if you can afford to have someone working on low-level things like that _full time_.